### DISCUSSION TITLE
importing non-manifold meshes

### DISCUSSION THREAD

**Discussion by Loosetooth** [elalish/manifold#471] [General]
Let's say that I would like to import an STL file in my app (could be triangle soup, and could also have some issues like non-manifoldness), and then I would like to do 3D boolean operations on them. It would be really nice to be able to use manifold for those operations, as it's really fast.

I gave this a try, and it works for most 'OK' STL files. However, when trying to construct the `Manifold` object from the meshdata, it sometimes complains about being unable to construct a `Manifold` type from the mesh because the result is non-manifold. Obviously, some of the STL files that I tried have problems.

The question: would it (hypothetically) be possible for manifold to handle incorrect meshdata, just to be able to do 3D booleans on them? I.e. we'd be able to do boolean operations, but the output would/could be non-manifold too, which would be OK in my use case.

**Comment by elalish** [comment]
There's more detailed discussion in our [wiki](https://github.com/elalish/manifold/wiki/Manifold-Library#manifoldness), but the short answer is no. A Boolean op is not defined for non-manifold meshes, because the winding number (inside vs. outside) is not defined for non-manifold meshes. Consider a mobius strip intersected with another mobius strip. What would that even mean? 

Generally when we're hoping to do Boolean ops on non-manifold meshes, they are meshes that "look" manifold - appear to represent a solid. In order for that Boolean op to be well defined and reliable, we need complete topological manifoldness. Any algorithm that doesn't require that is making guesses. Here we require you to do the guesswork ahead of time. Our `Merge` function is a simple version of this, but if your meshes are more broken than it can handle, you'll need to employ other manifold fixing algorithms, some of which are quite good (mostly for 3D printing), but are fundamentally heuristic.

**Comment by napter** [comment]
What libraries or algorithms are you aware of that can fix non-manifold
meshes and make them manifold? You said some are quite good, I'd love to
learn more.

**Reply by elalish** [reply]
I used to work with Netfabb, before they were acquired by Autodesk; they did an excellent job: https://www.autodesk.com/products/netfabb/free-trial

There's also MeshMixer (another defunct, but still free now-Autodesk product). I helped ship [3D Builder](https://apps.microsoft.com/store/detail/3d-builder/9WZDNCRFJ3T6?hl=en-us&gl=us) long ago, and it has great manifold fixing built-in (also powered by Netfabb), if you happen to have a Windows machine.

I would like to add support for proper manifold cleanup in this library eventually, but it'll be a bunch of work, and I'm not sure I want to deal with the volume of bugs, since there will always be cases it doesn't work well for.

**Comment by napter** [comment]
Thank you. Do you know of any open source libraries or algorithms that can
do this?

**Reply by elalish** [reply]
Not off the top of my head, but I'd look into the various open source Slicers for 3D printing, as they have to deal with this a lot.

**Reply by elalish** [reply]
Here are a bunch, though it's pretty out of date: http://meshrepair.org/

**Comment by pca006132** [comment]
I looked at CGAL, it seems that they implement the algorithm in "Cutting and Stitching: Converting Sets of Polygons to Manifold Surfaces", and it seems to work well. See https://github.com/openscad/openscad/pull/4832

**Reply by pca006132** [reply]
Note: only tested with trivial 2 cube cases (corner and edge touching). I don't have that many non-manifold test cases to test.

**Reply by elalish** [reply]
Non-manifold test cases are not generally hard to find - finding manifold meshes is much harder! Look for places where you can download 3D scans / photogrammetry - those are typically awful. Sketchup exports possibly the worst non-manifold geometry I've ever seen. Or just take a manifold mesh and randomly discard triangles.

**Reply by sloriot** [reply]
Note that you will end up with self-intersections where it is non-manifold.

**Reply by pca006132** [reply]
If I understand it correctly, those self-intersections caused by duplicated vertices are fine with our ε-valid definition?
